### PR TITLE
Queen of Hatred Code Improvement and Visual Changes

### PR DESCRIPTION
**Pull request by PotatoTomahto**
## About The Pull Request
Fixes:
- Fear being applied to dead bodies
- Arcana beats sigil not disappearing upon Qoh death
- Qoh can no longer be targeted or attacked while teleporting
- Player-controlled Qoh attacking when contained
- Qoh no longer depends on qliphoth counter when breached, only calls BreachEffect once
- Makes gift message include both "granted by ..." and unique message 

QoH Changes:
- Hostile breach inverts Qoh's colors when teleporting
- Teleport telefrag lowered and visually changed to match area of effect, no longer one-taps clerks (0 stats)
- Hostile beam has different charge up animation
- Qoh now teleports after every hostile beam
- Small change to records

3/4:
- Qoh checks for friendly fire so you don't get constantly blasted by her normal attacks

## Why It's Good For The Game
Fixes some minor bugs and makes QoH cooler and better. Make Qoh teleport around more will make her more evasive ~~and more annoying~~.
Improves the code so that Qoh will properly breach when friendly breaching. Attacks and other checks no longer check Qoh's qliphoth, which means you can properly spawn in a friendly Qoh with all her attacks available.
Changes gift message to provide information while also being unique.

## Changelog
:cl:
tweak: queen of hatred
:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 1]
```diff
@@ -139,10 +142,10 @@
 	return ..()
 
 /mob/living/simple_animal/hostile/abnormality/hatred_queen/AttackingTarget()
-	return OpenFire()
+	return OpenFire(target)
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:145]: This is not needed. Target is a simple mob variable, you don't actually have to put it in any proc most of the time.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 2]
```diff
@@ -139,10 +142,10 @@
...
 
 /mob/living/simple_animal/hostile/abnormality/hatred_queen/OpenFire()
-	if(!can_act)
+	if(!can_act || status_flags & GODMODE)
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:148]: This was previously handled by Para, so you don't have to put it here.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 3]
```diff
@@ -156,6 +159,8 @@
 				TryTeleport()
 			if(5)
 				if(datum_reference?.qliphoth_meter == 2) //only able to use normal if passive
+					if(prob(5))
+						addtimer(CALLBACK(src, .atom/movable/proc/say, "With love!"))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:163]: No. Don't force player controlled mobs to say stuff, it will annoy them.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 4]
```diff
@@ -167,7 +172,7 @@
 	if((beats_cooldown <= world.time) && prob(50))
 		ArcanaBeats(target)
 		return
-	if(prob(2))
+	if(prob(5))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:175]: Same here. 5% chance is high enough for it to get annoying.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 5]
```diff
@@ -226,27 +231,28 @@
...
+	spawned_effects += S
 	addtimer(CALLBACK(src, .atom/movable/proc/say, "Go! Arcana Beats~!"))
 	switch(dir)
 		if(EAST)
 			S.pixel_x += 16
 			var/matrix/new_matrix = matrix()
 			new_matrix.Scale(0.5, 1)
 			S.transform = new_matrix
-			S.layer += 0.1
+			S.layer += (src.layer + 0.1)
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:246]: Why?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 6]
```diff
@@ -292,36 +299,77 @@
...
+							new_matrix.Scale(0.5, 1)
+							S.transform = new_matrix
+							S.layer += (src.layer + 0.1)
+						if(SOUTH)
+							S.pixel_y += -20
+							S.layer += (src.layer + 0.1)
+						if(NORTH)
+							S.pixel_y += 28
+							S.layer -= 0.1
+			SLEEP_CHECK_DEATH(beam_startup) //time between beam startup stage
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:372]: All of it. Why did you do this?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 7]
```diff
@@ -344,22 +392,28 @@
 			if(accumulated_beam_damage >= 150)
 				if(beam_stage < 2)
 					beam_stage = 2
-					beam_damage_final *= 1.5
+					beam_damage_final *= 1.4
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:395]: Is that kind of a nerf even needed?

**Review by EgorDinamit — CHANGES_REQUESTED** [review]
There is way too many random edits and nerfs.
You might want to separate the changes so that I could review them better.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm (hunk 3 above)
Comment (Kitsunemitsu) [reply code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:163]: Just add a !client check

**Comment by Cupax3** [comment]
Gotta agree; More atomic PRs about your changes may be better.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 8]
```diff
@@ -139,35 +144,35 @@
...
 			if(3)
 				TryTeleport()
 			if(5)
-				if(datum_reference?.qliphoth_meter == 2) //only able to use normal if passive
+				if(friendly) //only able to use normal if passive
 					return ..()
 		return
 
-	if(beam_cooldown <= world.time && can_act && (prob(40) || datum_reference?.qliphoth_meter != 2)) //hostile breach should always be beaming
+	if(beam_cooldown <= world.time && (prob(40) || !friendly)) //hostile breach should always be beaming
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:167]: Why did you remove can_act check?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 9]
```diff
@@ -176,11 +181,14 @@
 	if(status_flags & GODMODE) // Contained
 		if(datum_reference?.qliphoth_meter == 1)
 			addtimer(CALLBACK(src, .proc/SpawnHeart), rand(4,10))
+	//if BREACHED, check if death_counter over the death limit
+	if(!(status_flags & GODMODE) && can_act && breach_max_death && (death_counter >= breach_max_death))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:185]: Why did you move it to Life()?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 10]
```diff
@@ -208,12 +216,10 @@
 	if(!died.ckey)
 		return FALSE
 	death_counter += 1
-	//if BREACHED, check if death_counter over the death limit
-	if(breach_max_death && (death_counter >= breach_max_death))
-		GoHysteric()
 	//if CONTAINED and not crazy
-	else if((status_flags & GODMODE) && (datum_reference?.qliphoth_meter == 2) && (death_counter > 3)) // Omagah a lot of dead people!
-		BreachEffect() // We must help them!
+	if((status_flags & GODMODE) && (datum_reference?.qliphoth_meter == 2) && (death_counter > 3)) // Omagah a lot of dead people!
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:220]: What are you even doing here? You've just made QoH hostile breach instead of breaching in friendly mode.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm (hunk 5 above)
Comment (EgorDinamit) [reply code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:247]: Now it will appear above pretty much everything.
Use `=` instead of `+=`

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 11]
```diff
@@ -290,55 +298,94 @@
...
+							S.transform = new_matrix
+							S.layer += 0.1
+						if(SOUTH)
+							S.pixel_y += -20
+							S.layer += (src.layer + 0.1)
+						if(NORTH)
+							S.pixel_y += 28
+							S.layer -= 0.1
+			SLEEP_CHECK_DEATH(beam_startup) //time between beam startup stage
+	var/turf/MT = friendly ? get_step(my_turf, my_dir) : get_turf(my_turf) //beam starts on the tile of qoh when hostile breach, allows stage 2 to hit people behind her
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:374]: Yeah, no, this is still not a good way to do it.

File: code/modules/mob/living/simple_animal/abnormality/_abnormality.dm [hunk 12]
```diff
@@ -356,8 +360,10 @@
 	datum_reference.abno_radio = TRUE
 
 /mob/living/simple_animal/hostile/abnormality/proc/IsContained() //Are you in a cell and currently contained?? If so stop.
-	if(status_flags & GODMODE)
+//Contained also checks if 1. qliphoth exists and 2. if qliphoth is not zero. Needed for when breached abnorms get set to godmode
+	if((status_flags & GODMODE) && (!datum_reference.qliphoth_meter_max || datum_reference.qliphoth_meter))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/_abnormality.dm:364]: This will cause unintended behavior with abnos like KoD, who do not have qliphoth counter but breach.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 13]
```diff
@@ -139,63 +140,67 @@
...
 		BeamAttack(target)
 		return
-	if(!("neutral" in faction))
+	if(!friendly)
 		return
 	if((beats_cooldown <= world.time) && prob(50))
 		ArcanaBeats(target)
 		return
-	if(prob(2))
+	if(prob(4))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:171]: Will certainly get annoying...

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 14]
```diff
@@ -139,63 +140,67 @@
...
 
 /mob/living/simple_animal/hostile/abnormality/hatred_queen/Life()
 	. = ..()
-	if(status_flags & GODMODE) // Contained
+	if(IsContained()) // Contained
 		if(datum_reference?.qliphoth_meter == 1)
+			addtimer(CALLBACK(src, .proc/SpawnHeart), rand(2,8))
 			addtimer(CALLBACK(src, .proc/SpawnHeart), rand(4,10))
+	//if BREACHED, check if death_counter over the death limit
+	if(!IsContained() && can_act && breach_max_death && (death_counter >= breach_max_death))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:182]: Again, why are you moving it to Life() proc?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 15]
```diff
@@ -205,15 +210,13 @@
...
+	if(!died.mind)
 		return FALSE
 	death_counter += 1
-	//if BREACHED, check if death_counter over the death limit
-	if(breach_max_death && (death_counter >= breach_max_death))
-		GoHysteric()
 	//if CONTAINED and not crazy
-	else if((status_flags & GODMODE) && (datum_reference?.qliphoth_meter == 2) && (death_counter > 3)) // Omagah a lot of dead people!
-		BreachEffect() // We must help them!
+	if(IsContained() && friendly && (death_counter > 3)) // Omagah a lot of dead people!
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:217]: Why are you reverting the behavior?

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 16]
```diff
@@ -23,6 +24,8 @@
 	del_on_death = FALSE
 	projectilesound = 'sound/abnormalities/hatredqueen/attack.ogg'
 	deathsound = 'sound/abnormalities/hatredqueen/dead.ogg'
+	deathmessage = "slowly falls to the ground."
+	check_friendly_fire = 1
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:28]:
```suggestion
	check_friendly_fire = TRUE
```
Don't use 1/0 for booleans.

File: code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm [hunk 17]
```diff
@@ -417,37 +468,53 @@
...
-		if(L.stat == DEAD || L.z != z || L.status_flags & GODMODE)
+		if(L.stat == DEAD || L.z != z || L.status_flags & GODMODE || faction_check_mob(L))
 			continue
-		if(datum_reference?.qliphoth_meter != 2 && !ishuman(L))
+		if(!friendly && !ishuman(L))
 			continue
-		var/turf/T = get_turf(L)
-		if(!faction_check_mob(L) && L.stat != DEAD)
-			teleport_potential += T
+		if(ishuman(L))
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/waw/hatred_queen.dm:475]: Obsolete check. We're ignoring non-humans due to check above already.
